File: 061_HouseRobber.cpp
// Topic: Dynamic Programming
// House Robber: maximize sum of non-adjacent houses using O(1) space DP
// Time Complexity: O(n)
// Space Complexity: O(1)

#include <bits/stdc++.h>
using namespace std;

int rob(vector<int>& nums)
{
    int incl = 0, excl = 0;
    for (int x : nums)
    {
        int new_excl = max(incl, excl);
        incl = excl + x;
        excl = new_excl;
    }
    return max(incl, excl);
}

int main()
{
    cout << rob((vector<int>{1, 2, 3, 1})) << "\n";     // 4
    cout << rob((vector<int>{2, 7, 9, 3, 1})) << "\n";  // 12
    return 0;
}

File: 061_HouseRobber.py
"""
Topic: Dynamic Programming
Problem: House Robber
Description: Max sum of non-adjacent elements.
Time: O(n)
Space: O(1)
"""

from typing import List


def rob(nums: List[int]) -> int:
    prev = 0
    curr = 0
    for x in nums:
        prev, curr = curr, max(curr, prev + x)
    return curr


if __name__ == '__main__':
    print(rob([1,2,3,1]))  # 4
    print(rob([2,7,9,3,1]))  # 12
